cpp/include/cuspatial/detail/utility/traits.hpp

```diff
@@ -23,5 +23,44 @@ constexpr bool is_floating_point()
   return std::conjunction_v<std::is_floating_point<Ts>...>;
 }
 
+/**
+ * @internal
+ * @brief returns true if all types are floating point types.
+ */
+template <typename... Ts>
+constexpr bool is_integral()
+{
+  return std::conjunction_v<std::is_integral<Ts>...>;
+}
+
+/**
+ * @internal
+ * @brief returns true if T and all types Ts... are the same floating point type.
+ */
+template <typename T, typename... Ts>
+constexpr bool is_same_floating_point()
+{
+  return std::conjunction_v<std::is_same<T, Ts>...> and
+         std::conjunction_v<std::is_floating_point<Ts>...>;
+}
```

Comment:
I thought it would be because of the header only libraries. We're just generating an error for them at compile time to force them to write the correct types, instead of finding out what happens.... later....

Reply:
Where "them" may be "us" when talking about the layers on top of the header-only-API.